PIGPIO/pigpio.3

." Process this file with
." groff -man -Tascii pigpio.3
."
.TH pigpio 3 2012-2015 Linux "pigpio archive"
.SH NAME
pigpio - A C library to manipulate the Pi's GPIO.

.SH SYNOPSIS

#include <pigpio.h>


gcc -Wall -pthread -o prog prog.c -lpigpio -lrt

sudo ./prog
.SH DESCRIPTION


.ad l

.nh

.br

.br
pigpio is a C library for the Raspberry which allows control of the GPIO.

.br

.br
.SS Features
.br

.br
o hardware timed PWM on any of GPIO 0-31

.br

.br
o hardware timed servo pulses on any of GPIO 0-31

.br

.br
o callbacks when any of GPIO 0-31 change state

.br

.br
o callbacks at timed intervals

.br

.br
o reading/writing all of the GPIO in a bank as one operation

.br

.br
o individually setting GPIO modes, reading and writing

.br

.br
o notifications when any of GPIO 0-31 change state

.br

.br
o the construction of output waveforms with microsecond timing

.br

.br
o rudimentary permission control over GPIO

.br

.br
o a simple interface to start and stop new threads

.br

.br
o I2C, SPI, and serial link wrappers

.br

.br
o creating and running scripts

.br

.br
.SS GPIO
.br

.br
ALL GPIO are identified by their Broadcom number.

.br

.br
.SS Credits
.br

.br
The PWM and servo pulses are timed using the DMA and PWM peripherals.

.br

.br
This use was inspired by Richard Hirst's servoblaster kernel module.

.br

.br
See \fBhttps://github.com/richardghirst/PiBits/tree/master/ServoBlaster\fP

.br

.br
.SS Usage
.br

.br
Include <pigpio.h> in your source files.

.br

.br
Assuming your source is in prog.c use the following command to build and
run the executable.

.br

.br

.EX
gcc -Wall -pthread -o prog prog.c -lpigpio -lrt
.br
sudo ./prog
.br

.EE

.br

.br
For examples of usage see the C programs within the pigpio archive file.

.br

.br
.SS Notes
.br

.br
All the functions which return an int return < 0 on error.

.br

.br
\fBgpioInitialise\fP must be called before all other library functions
with the following exceptions:

.br

.br

.EX
\fBgpioCfg*\fP
.br
\fBgpioVersion\fP
.br
\fBgpioHardwareRevision\fP
.br

.EE

.br

.br
If the library is not initialised all but the \fBgpioCfg*\fP,
\fBgpioVersion\fP, and \fBgpioHardwareRevision\fP functions will
return error PI_NOT_INITIALISED.

.br

.br
If the library is initialised the \fBgpioCfg*\fP functions will return
error PI_INITIALISED.

.br

.br
.SH FUNCTIONS

.IP "\fBint gpioInitialise(void)\fP"
.IP "" 4
Initialises the library.

.br

.br
Returns the pigpio version number if OK, otherwise PI_INIT_FAILED.

.br

.br
gpioInitialise must be called before using the other library functions
with the following exceptions:

.br

.br

.EX
\fBgpioCfg*\fP
.br
\fBgpioVersion\fP
.br
\fBgpioHardwareRevision\fP
.br

.EE

.br

.br
\fBExample\fP
.br

.EX
if (gpioInitialise() < 0)
.br
{
.br
   // pigpio initialisation failed.
.br
}
.br
else
.br
{
.br
   // pigpio initialised okay.
.br
}
.br

.EE

.IP "\fBvoid gpioTerminate(void)\fP"
.IP "" 4
Terminates the library.

.br

.br
Returns nothing.

.br

.br
Call before program exit.

.br

.br
This function resets the used DMA channels, releases memory, and
terminates any running threads.

.br

.br
\fBExample\fP
.br

.EX
gpioTerminate();
.br

.EE

.IP "\fBint gpioSetMode(unsigned gpio, unsigned mode)\fP"
.IP "" 4
Sets the GPIO mode, typically input or output.

.br

.br

.EX
gpio: 0-53
.br
mode: 0-7
.br

.EE

.br

.br
Returns 0 if OK, otherwise PI_BAD_GPIO or PI_BAD_MODE.

.br

.br
Arduino style: pinMode.

.br

.br
\fBExample\fP
.br

.EX
gpioSetMode(17, PI_INPUT);  // Set GPIO17 as input.
.br

.br
gpioSetMode(18, PI_OUTPUT); // Set GPIO18 as output.
.br

.br
gpioSetMode(22,PI_ALT0);    // Set GPIO22 to alternative mode 0.
.br

.EE

.IP "\fBint gpioGetMode(unsigned gpio)\fP"
.IP "" 4
Gets the GPIO mode.

.br

.br

.EX
gpio: 0-53
.br

.EE

.br

.br
Returns the GPIO mode if OK, otherwise PI_BAD_GPIO.

.br

.br
\fBExample\fP
.br

.EX
if (gpioGetMode(17) != PI_ALT0)
.br
{
.br
   gpioSetMode(17, PI_ALT0);  // set GPIO17 to ALT0
.br
}
.br

.EE

.IP "\fBint gpioSetPullUpDown(unsigned gpio, unsigned pud)\fP"
.IP "" 4
Sets or clears resistor pull ups or downs on the GPIO.

.br

.br

.EX
gpio: 0-53
.br
 pud: 0-2
.br

.EE

.br

.br
Returns 0 if OK, otherwise PI_BAD_GPIO or PI_BAD_PUD.

.br

.br
\fBExample\fP
.br

.EX
gpioSetPullUpDown(17, PI_PUD_UP);   // Sets a pull-up.
.br

.br
gpioSetPullUpDown(18, PI_PUD_DOWN); // Sets a pull-down.
.br

.br
gpioSetPullUpDown(23, PI_PUD_OFF);  // Clear any pull-ups/downs.
.br

.EE

.IP "\fBint gpioRead(unsigned gpio)\fP"
.IP "" 4
Reads the GPIO level, on or off.

.br

.br

.EX
gpio: 0-53
.br

.EE

.br

.br
Returns the GPIO level if OK, otherwise PI_BAD_GPIO.

.br

.br
Arduino style: digitalRead.

.br

.br
\fBExample\fP
.br

.EX
printf("GPIO24 is level %d", gpioRead(24));
.br

.EE

.IP "\fBint gpioWrite(unsigned gpio, unsigned level)\fP"
.IP "" 4
Sets the GPIO level, on or off.

.br

.br

.EX
 gpio: 0-53
.br
level: 0-1
.br

.EE

.br

.br
Returns 0 if OK, otherwise PI_BAD_GPIO or PI_BAD_LEVEL.

.br

.br
If PWM or servo pulses are active on the GPIO they are switched off.

.br

.br
Arduino style: digitalWrite

.br

.br
\fBExample\fP
.br

.EX
gpioWrite(24, 1); // Set GPIO24 high.
.br

.EE

.IP "\fBint gpioPWM(unsigned user_gpio, unsigned dutycycle)\fP"
.IP "" 4
Starts PWM on the GPIO, dutycycle between 0 (off) and range (fully on).
Range defaults to 255.

.br

.br

.EX
user_gpio: 0-31
.br
dutycycle: 0-range
.br

.EE

.br

.br
Returns 0 if OK, otherwise PI_BAD_USER_GPIO or PI_BAD_DUTYCYCLE.

.br

.br
Arduino style: analogWrite

.br

.br
This and the servo functionality use the DMA and PWM or PCM peripherals
to control and schedule the pulse lengths and dutycycles.

.br

.br
The \fBgpioSetPWMrange\fP function may be used to change the default
range of 255.

.br

.br
\fBExample\fP
.br

.EX
gpioPWM(17, 255); // Sets GPIO17 full on.
.br

.br
gpioPWM(18, 128); // Sets GPIO18 half on.
.br

.br
gpioPWM(23, 0);   // Sets GPIO23 full off.
.br

.EE

.IP "\fBint gpioGetPWMdutycycle(unsigned user_gpio)\fP"
.IP "" 4
Returns the PWM dutycycle setting for the GPIO.

.br

.br

.EX
user_gpio: 0-31
.br

.EE

.br

.br
Returns between 0 (off) and range (fully on) if OK, otherwise
PI_BAD_USER_GPIO or PI_NOT_PWM_GPIO.

.br

.br
For normal PWM the dutycycle will be out of the defined range
for the GPIO (see \fBgpioGetPWMrange\fP).

.br

.br
If a hardware clock is active on the GPIO the reported dutycycle
will be 500000 (500k) out of 1000000 (1M).

.br

.br
If hardware PWM is active on the GPIO the reported dutycycle
will be out of a 1000000 (1M).

.br

.br
Normal PWM range defaults to 255.

.IP "\fBint gpioSetPWMrange(unsigned user_gpio, unsigned range)\fP"
.IP "" 4
Selects the dutycycle range to be used for the GPIO.  Subsequent calls
to gpioPWM will use a dutycycle between 0 (off) and range (fully on).

.br

.br

.EX
user_gpio: 0-31
.br
    range: 25-40000
.br

.EE

.br

.br
Returns the real range for the given GPIO's frequency if OK,
otherwise PI_BAD_USER_GPIO or PI_BAD_DUTYRANGE.

.br

.br
If PWM is currently active on the GPIO its dutycycle will be scaled
to reflect the new range.

.br

.br
The real range, the number of steps between fully off and fully
on for each frequency, is given in the following table.

.br

.br

.EX
  25,   50,  100,  125,  200,  250,  400,   500,   625,
.br
 800, 1000, 1250, 2000, 2500, 4000, 5000, 10000, 20000
.br

.EE

.br

.br
The real value set by \fBgpioPWM\fP is (dutycycle * real range) / range.

.br

.br
\fBExample\fP
.br

.EX
gpioSetPWMrange(24, 2000); // Now 2000 is fully on
.br
                           //     1000 is half on
.br
                           //      500 is quarter on, etc.
.br

.EE

.IP "\fBint gpioGetPWMrange(unsigned user_gpio)\fP"
.IP "" 4
Returns the dutycycle range used for the GPIO if OK, otherwise
PI_BAD_USER_GPIO.

.br

.br

.EX
user_gpio: 0-31
.br

.EE

.br

.br
If a hardware clock or hardware PWM is active on the GPIO
the reported range will be 1000000 (1M).

.br

.br
\fBExample\fP
.br

.EX
r = gpioGetPWMrange(23);
.br

.EE

.IP "\fBint gpioGetPWMrealRange(unsigned user_gpio)\fP"
.IP "" 4
Returns the real range used for the GPIO if OK, otherwise
PI_BAD_USER_GPIO.

.br

.br

.EX
user_gpio: 0-31
.br

.EE

.br

.br
If a hardware clock is active on the GPIO the reported real
range will be 1000000 (1M).

.br

.br
If hardware PWM is active on the GPIO the reported real range
will be approximately 250M divided by the set PWM frequency.

.br

.br
\fBExample\fP
.br

.EX
rr = gpioGetPWMrealRange(17);
.br

.EE

.IP "\fBint gpioSetPWMfrequency(unsigned user_gpio, unsigned frequency)\fP"
.IP "" 4
Sets the frequency in hertz to be used for the GPIO.

.br

.br

.EX
user_gpio: 0-31
.br
frequency: >=0
.br

.EE

.br

.br
Returns the numerically closest frequency if OK, otherwise
PI_BAD_USER_GPIO.

.br

.br
If PWM is currently active on the GPIO it will be
switched off and then back on at the new frequency.

.br

.br
Each GPIO can be independently set to one of 18 different PWM
frequencies.

.br

.br
The selectable frequencies depend upon the sample rate which
may be 1, 2, 4, 5, 8, or 10 microseconds (default 5).

.br

.br
The frequencies for each sample rate are:

.br

.br

.EX
                       Hertz
.br

.br
       1: 40000 20000 10000 8000 5000 4000 2500 2000 1600
.br
           1250  1000   800  500  400  250  200  100   50
.br

.br
       2: 20000 10000  5000 4000 2500 2000 1250 1000  800
.br
            625   500   400  250  200  125  100   50   25
.br

.br
       4: 10000  5000  2500 2000 1250 1000  625  500  400
.br
            313   250   200  125  100   63   50   25   13
.br
sample
.br
 rate
.br
 (us)  5:  8000  4000  2000 1600 1000  800  500  400  320
.br
            250   200   160  100   80   50   40   20   10
.br

.br
       8:  5000  2500  1250 1000  625  500  313  250  200
.br
            156   125   100   63   50   31   25   13    6
.br

.br
      10:  4000  2000  1000  800  500  400  250  200  160
.br
            125   100    80   50   40   25   20   10    5
.br

.EE

.br

.br
\fBExample\fP
.br

.EX
gpioSetPWMfrequency(23, 0); // Set GPIO23 to lowest frequency.
.br

.br
gpioSetPWMfrequency(24, 500); // Set GPIO24 to 500Hz.
.br

.br
gpioSetPWMfrequency(25, 100000); // Set GPIO25 to highest frequency.
.br

.EE

.IP "\fBint gpioGetPWMfrequency(unsigned user_gpio)\fP"
.IP "" 4
Returns the frequency (in hertz) used for the GPIO if OK, otherwise
PI_BAD_USER_GPIO.

.br

.br

.EX
user_gpio: 0-31
.br

.EE

.br

.br
For normal PWM the frequency will be that defined for the GPIO by
\fBgpioSetPWMfrequency\fP.

.br

.br
If a hardware clock is active on the GPIO the reported frequency
will be that set by \fBgpioHardwareClock\fP.

.br

.br
If hardware PWM is active on the GPIO the reported frequency
will be that set by \fBgpioHardwarePWM\fP.

.br

.br
\fBExample\fP
.br

.EX
f = gpioGetPWMfrequency(23); // Get frequency used for GPIO23.
.br

.EE

.IP "\fBint gpioServo(unsigned user_gpio, unsigned pulsewidth)\fP"
.IP "" 4
Starts servo pulses on the GPIO, 0 (off), 500 (most anti-clockwise) to
2500 (most clockwise).

.br

.br

.EX
 user_gpio: 0-31
.br
pulsewidth: 0, 500-2500
.br

.EE

.br

.br
Returns 0 if OK, otherwise PI_BAD_USER_GPIO or PI_BAD_PULSEWIDTH.

.br

.br
The range supported by servos varies and should probably be determined
by experiment.  A value of 1500 should always be safe and represents
the mid-point of rotation.  You can DAMAGE a servo if you command it
to move beyond its limits.

.br

.br
The following causes an on pulse of 1500 microseconds duration to be
transmitted on GPIO 17 at a rate of 50 times per second. This will
command a servo connected to GPIO 17 to rotate to its mid-point.

.br

.br
\fBExample\fP
.br

.EX
gpioServo(17, 1000); // Move servo to safe position anti-clockwise.
.br

.br
gpioServo(23, 1500); // Move servo to centre position.
.br

.br
gpioServo(25, 2000); // Move servo to safe position clockwise.
.br

.EE

.br

.br
OTHER UPDATE RATES:

.br

.br
This function updates servos at 50Hz.  If you wish to use a different
update frequency you will have to use the PWM functions.

.br

.br

.EX
PWM Hz    50   100  200  400  500
.br
1E6/Hz 20000 10000 5000 2500 2000
.br

.EE

.br

.br
Firstly set the desired PWM frequency using \fBgpioSetPWMfrequency\fP.

.br

.br
Then set the PWM range using \fBgpioSetPWMrange\fP to 1E6/frequency.
Doing this allows you to use units of microseconds when setting
the servo pulsewidth.

.br

.br
E.g. If you want to update a servo connected to GPIO25 at 400Hz

.br

.br

.EX
gpioSetPWMfrequency(25, 400);
.br

.br
gpioSetPWMrange(25, 2500);
.br

.EE

.br

.br
Thereafter use the PWM command to move the servo,
e.g. gpioPWM(25, 1500) will set a 1500 us pulse.

.IP "\fBint gpioGetServoPulsewidth(unsigned user_gpio)\fP"
.IP "" 4
Returns the servo pulsewidth setting for the GPIO.

.br

.br

.EX
user_gpio: 0-31
.br

.EE

.br

.br
Returns 0 (off), 500 (most anti-clockwise) to 2500 (most clockwise)
if OK, otherwise PI_BAD_USER_GPIO or PI_NOT_SERVO_GPIO.

.IP "\fBint gpioSetAlertFunc(unsigned user_gpio, gpioAlertFunc_t f)\fP"
.IP "" 4
Registers a function to be called (a callback) when the specified
GPIO changes state.

.br

.br

.EX
user_gpio: 0-31
.br
        f: the callback function
.br

.EE

.br

.br
Returns 0 if OK, otherwise PI_BAD_USER_GPIO.

.br

.br
One function may be registered per GPIO.

.br

.br
The function is passed the GPIO, the new level, and the tick.

.br

.br
The alert may be cancelled by passing NULL as the function.

.br

.br
The GPIO are sampled at a rate set when the library is started.

.br

.br
If a value isn't specifically set the default of 5 us is used.

.br

.br
The number of samples per second is given in the following table.

.br

.br

.EX
              samples
.br
              per sec
.br

.br
         1  1,000,000
.br
         2    500,000
.br
sample   4    250,000
.br
rate     5    200,000
.br
(us)     8    125,000
.br
        10    100,000
.br

.EE

.br

.br
Level changes shorter than the sample rate may be missed.

.br

.br
The thread which calls the alert functions is triggered nominally
1000 times per second.  The active alert functions will be called
once per level change since the last time the thread was activated.
i.e. The active alert functions will get all level changes but there
will be a latency.

.br

.br
The tick value is the time stamp of the sample in microseconds, see
\fBgpioTick\fP for more details.

.br

.br
\fBExample\fP
.br

.EX
void aFunction(int gpio, int level, uint32_t tick)
.br
{
.br
   printf("GPIO %d became %d at %d", gpio, level, tick);
.br
}
.br

.br
// call aFunction whenever GPIO 4 changes state
.br

.br
gpioSetAlertFunc(4, aFunction);
.br

.EE

.IP "\fBint gpioSetAlertFuncEx(unsigned user_gpio, gpioAlertFuncEx_t f, void *userdata)\fP"
.IP "" 4
Registers a function to be called (a callback) when the specified
GPIO changes state.

.br

.br

.EX
user_gpio: 0-31
.br
        f: the callback function
.br
 userdata: pointer to arbitrary user data
.br

.EE

.br

.br
Returns 0 if OK, otherwise PI_BAD_USER_GPIO.

.br

.br
One function may be registered per GPIO.

.br

.br
The function is passed the GPIO, the new level, the tick, and
the userdata pointer.

.br

.br
Only one of \fBgpioSetAlertFunc\fP or \fBgpioSetAlertFuncEx\fP can be
registered per GPIO.

.br

.br
See \fBgpioSetAlertFunc\fP for further details.

.IP "\fBint gpioSetISRFunc(unsigned user_gpio, unsigned edge, int timeout, gpioISRFunc_t f)\fP"
.IP "" 4
Registers a function to be called (a callback) whenever the specified
GPIO interrupt occurs.

.br

.br

.EX
user_gpio: 0-31
.br
     edge: RISING_EDGE, FALLING_EDGE, or EITHER_EDGE
.br
  timeout: interrupt timeout in milliseconds (<=0 to cancel)
.br
        f: the callback function
.br

.EE

.br

.br
Returns 0 if OK, otherwise PI_BAD_USER_GPIO, PI_BAD_EDGE,
or PI_BAD_ISR_INIT.

.br

.br
One function may be registered per GPIO.

.br

.br
The function is passed the GPIO, the current level, and the
current tick.  The level will be PI_TIMEOUT if the optional
interrupt timeout expires.

.br

.br
The underlying Linux sysfs GPIO interface is used to provide
the interrupt services.

.br

.br
The first time the function is called, with a non-NULL f, the
GPIO is exported, set to be an input, and set to interrupt
on the given edge and timeout.

.br

.br
Subsequent calls, with a non-NULL f, can vary one or more of the
edge, timeout, or function.

.br

.br
The ISR may be cancelled by passing a NULL f, in which case the
GPIO is unexported.

.br

.br
The tick is that read at the time the process was informed of
the interrupt.  This will be a variable number of microseconds
after the interrupt occurred.  Typically the latency will be of
the order of 50 microseconds.  The latency is not guaranteed
and will vary with system load.

.br

.br
The level is that read at the time the process was informed of
the interrupt, or PI_TIMEOUT if the optional interrupt timeout
expired.  It may not be the same as the expected edge as
interrupts happening in rapid succession may be missed by the
kernel (i.e. this mechanism can not be used to capture several
interrupts only a few microseconds apart).

.IP "\fBint gpioSetISRFuncEx(unsigned user_gpio, unsigned edge, int timeout, gpioISRFuncEx_t f, void *userdata)\fP"
.IP "" 4
Registers a function to be called (a callback) whenever the specified
GPIO interrupt occurs.

.br

.br

.EX
user_gpio: 0-31
.br
     edge: RISING_EDGE, FALLING_EDGE, or EITHER_EDGE
.br
  timeout: interrupt timeout in milliseconds (<=0 to cancel)
.br
        f: the callback function
.br
 userdata: pointer to arbitrary user data
.br

.EE

.br

.br
Returns 0 if OK, otherwise PI_BAD_USER_GPIO, PI_BAD_EDGE,
or PI_BAD_ISR_INIT.

.br

.br
The function is passed the GPIO, the current level, the
current tick, and the userdata pointer.

.br

.br
Only one of \fBgpioSetISRFunc\fP or \fBgpioSetISRFuncEx\fP can be
registered per GPIO.

.br

.br
See \fBgpioSetISRFunc\fP for further details.

.IP "\fBint gpioNotifyOpen(void)\fP"
.IP "" 4
This function requests a free notification handle.

.br

.br
Returns a handle greater than or equal to zero if OK,
otherwise PI_NO_HANDLE.

.br

.br
A notification is a method for being notified of GPIO state changes
via a pipe or socket.

.br

.br
Pipe notifications for handle x will be available at the pipe
named /dev/pigpiox (where x is the handle number).  E.g. if the
function returns 15 then the notifications must be read
from /dev/pigpio15.

.br

.br
Socket notifications are returned to the socket which requested the
handle.

.br

.br
\fBExample\fP
.br

.EX
h = gpioNotifyOpen();
.br

.br
if (h >= 0)
.br
{
.br
   sprintf(str, "/dev/pigpio%d", h);
.br

.br
   fd = open(str, O_RDONLY);
.br

.br
   if (fd >= 0)
.br
   {
.br
      // Okay.
.br
   }
.br
   else
.br
   {
.br
      // Error.
.br
   }
.br
}
.br
else
.br
{
.br
   // Error.
.br
}
.br

.EE

.IP "\fBint gpioNotifyOpenWithSize(int bufSize)\fP"
.IP "" 4
This function requests a free notification handle.

.br

.br
It differs from \fBgpioNotifyOpen\fP in that the pipe size may be
specified, whereas \fBgpioNotifyOpen\fP uses the default pipe size.

.br

.br
See \fBgpioNotifyOpen\fP for further details.

.IP "\fBint gpioNotifyBegin(unsigned handle, uint32_t bits)\fP"
.IP "" 4
This function starts notifications on a previously opened handle.

.br

.br

.EX
handle: >=0, as returned by \fBgpioNotifyOpen\fP
.br
  bits: a bit mask indicating the GPIO of interest
.br

.EE

.br

.br
Returns 0 if OK, otherwise PI_BAD_HANDLE.

.br

.br
The notification sends state changes for each GPIO whose corresponding
bit in bits is set.

.br

.br
Each notification occupies 12 bytes in the fifo and has the
following structure.

.br

.br

.EX
typedef struct
.br
{
.br
   uint16_t seqno;
.br
   uint16_t flags;
.br
   uint32_t tick;
.br
   uint32_t level;
.br
} gpioReport_t;
.br

.EE

.br

.br
seqno: starts at 0 each time the handle is opened and then increments
by one for each report.

.br

.br
flags: three flags are defined, PI_NTFY_FLAGS_WDOG,
PI_NTFY_FLAGS_ALIVE, and PI_NTFY_FLAGS_EVENT.

.br

.br
If bit 5 is set (PI_NTFY_FLAGS_WDOG) then bits 0-4 of the flags
indicate a GPIO which has had a watchdog timeout.

.br

.br
If bit 6 is set (PI_NTFY_FLAGS_ALIVE) this indicates a keep alive
signal on the pipe/socket and is sent once a minute in the absence
of other notification activity.

.br

.br
If bit 7 is set (PI_NTFY_FLAGS_EVENT) then bits 0-4 of the flags
indicate an event which has been triggered.

.br

.br
tick: the number of microseconds since system boot.  It wraps around
after 1h12m.

.br

.br
level: indicates the level of each GPIO.  If bit 1<<x is set then
GPIO x is high.

.br

.br
\fBExample\fP
.br

.EX
// Start notifications for GPIO 1, 4, 6, 7, 10.
.br

.br

CHANGELOG.md
<a name="1.0.0"></a>
# [1.0.0](https://github.com/i-am-digital/js-gpiozero/compare/1.0.0...v1.0.0) (2016-12-29)




<a name="0.1.5"></a>
## [0.1.5](https://github.com/i-am-digital/js-gpiozero/compare/0.1.5...v0.1.5) (2016-12-27)






gpiozero/pins/pigpio.js
const pigpio = require('js-pigpio');
const p = require("./index.js");
const inherit = require('../tools.js').inherit;
const Pin = require("./index.js").Pin;
const PiBoardInfo = require("./data.js").PiBoardInfo;

const _CONNECTIONS = []; // maps (host, port) to (connection, pi_info)
const _PINS = [];
/*
const GPIO_EDGES = {
    'both':    pigpio.EITHER_EDGE,
    'rising':  pigpio.RISING_EDGE,
    'falling': pigpio.FALLING_EDGE,
}*/

const GPIO_PULL_UPS = {
    'up':       pigpio.PUD_UP,
    'down':     pigpio.PUD_DOWN,
    'floating': pigpio.PUD_OFF,
};

const GPIO_FUNCTIONS = {
    'input':   pigpio.INPUT,
    'output':  pigpio.OUTPUT,
    'alt0':    pigpio.ALT0,
    'alt1':    pigpio.ALT1,
    'alt2':    pigpio.ALT2,
    'alt3':    pigpio.ALT3,
    'alt4':    pigpio.ALT4,
    'alt5':    pigpio.ALT5,
};

const GPIO_FUNCTION_NAMES = {};
for(var name in GPIO_FUNCTIONS) {
    var value = GPIO_FUNCTIONS[name];
    GPIO_FUNCTIONS[value] = GPIO_FUNCTIONS[name];
}
//GPIO_FUNCTION_NAMES = {v: k for (k, v) in GPIO_FUNCTIONS.items()}

/**
 *
 * Uses the library {@link https://www.npmjs.com/package/js-pigpio|js-pigpio} to interface to the Pi's GPIO pins. The pigpio library relies on a daemon {@link http://abyz.co.uk/rpi/pigpio/|pigpiod}  to be running as root to provide access to the GPIO pins, and communicates with this daemon over a network socket.
 *
 * While this does mean only the daemon itself should control the pins, the architecture does have several advantages:
 * * Pins can be remote controlled from another machine (the other machine doesn't even have to be a Raspberry Pi; it simply needs the `pigpio`_ client library installed on it),
 * * The daemon supports hardware PWM via the DMA controller.
 * * Your script itself doesn't require root privileges; it just needs to be able to communicate with the daemon.
 *
 * @example
 * from gpiozero.pins.pigpiod import PiGPIOPin
 * from gpiozero import LED
 * led = LED(PiGPIOPin(12))
 *
 * This is particularly useful for controlling pins on a remote machine. To
 * accomplish this simply specify the host (and optionally port) when
 * constructing the pin.
 *
 * @example
 * from gpiozero.pins.pigpiod import PiGPIOPin
 * from gpiozero import LED
 * from signal import pause
 * led = LED(PiGPIOPin(12, '192.168.0.2',5000)
 *
 * @param {number} number - GPIO Pin.
 * @param {string} [host] - IP address of remote server defaults to localhost.
 * @param {string} [port] - Port of remote server.
 * @returns {PiGPIOPin} - An instance of a PiGPIOPin.
 *
 * @class
 */
function PiGPIOPin (number, host, port) {
    if (host === undefined) {
        host = process.env.PIGPIO_ADDR || 'localhost';
    }
    if (port === undefined) {
        port =  process.env.PIGPIO_PORT || '8888';
    }
    const old_pin = _PINS[number+':'+host+':'+port];
    if (old_pin === undefined) {
        Pin.call(this);

        this._number = number;
        this._host = host;
        this._port = port;
        this._function = 'input';
        this._state = false;
        this._bounce = null;
        this._edges =  pigpio.EITHER_EDGE;
        this._when_changed = null;
        this._callback = null;
        this.pi_info(host, port);
        _PINS[number+':'+host+':'+port] = this;
        return this;
    }
}

PiGPIOPin.prototype = inherit(p.Pin.prototype);
PiGPIOPin.prototype.constructor = PiGPIOPin;

exports.PiGPIOPin = PiGPIOPin;

PiGPIOPin.prototype.pi_info = function (host, port) {
    if (host === undefined) {
        host = process.env.PIGPIO_ADDR || 'localhost';
    }
    if (port === undefined) {
        port =  process.env.PIGPIO_PORT || '8888';
    }
    let connection = _CONNECTIONS[host+':'+ port];

    if(connection === undefined) {
        const that = this;
        connection = pigpio.pi(host, port, (err) => {
            if (err !== undefined) {
                throw new Error(err);
            }
            connection.get_hardware_revision((err,data)=> {
                const info = PiBoardInfo(data);
                that._connection = connection;
                that._pi_info = info;
                that._pi_info.physical_pin('GPIO' + that._number); //Cause an error if not valid
                that._pull =  that._pi_info.pulled_up('GPIO' + that._number)?'up':'floating';
                that._connection.set_mode(this._number, pigpio.INPUT);
                that._connection.set_pull_up_down(that._number, GPIO_PULL_UPS[this._pull]);
                that._connection.set_glitch_filter(that._number, 0);
                _CONNECTIONS[host + ':' + port] = {connection, info};
            });
        });
    } else {
        this._connection = connection.connection;
        this._pi_info = connection._info;
    }
};

/**
 * If we're shutting down, the connection may have disconnected itself
 * already. Unfortunately, the connection's "connected" property is
 * rather buggy - disconnecting doesn't set it to False! So we're
 * naughty and check an internal variable instead...
 */
PiGPIOPin.prototype.close = function () {
    if (this._connection.sl.s !== undefined) {
        this.frequency(null);
        this.when_changed(null);
        this.pin_function('input')
        this.pull('up');
        if (this._pi_info.pulled_up('GPIO'+this.number)) {
            this.pull('up');
        } else {
            this.pull('floating');
        }
    }
};

/*
 def _get_function(self):
 return self.GPIO_FUNCTION_NAMES[self._connection.get_mode(self._number)]

 def _set_function(self, value):
 if value != 'input':
 self._pull = 'floating'
 try:
 self._connection.set_mode(self._number, self.GPIO_FUNCTIONS[value])
 except KeyError:
 raise PinInvalidFunction('invalid function "%s" for pin %r' % (value, self))

 */

PiGPIOPin.prototype.pin_function = function (value) {
    // need to check if this is a callback function
    if (value === undefined) {
        const callback = value;
        return GPIO_FUNCTION_NAMES[this._connection.get_mode(this._number, callback)];
    }
    if(value !== 'input') {
        this._pull = 'floating';
    }
    this._connection.set_mode(this._number, GPIO_FUNCTIONS[value]);
}

 /*

GPIO_FUNCTION_NAMES = {v: k for (k, v) in GPIO_FUNCTIONS.items()}
GPIO_PULL_UP_NAMES = {v: k for (k, v) in GPIO_PULL_UPS.items()}
GPIO_EDGES_NAMES = {v: k for (k, v) in GPIO_EDGES.items()}

def __repr__(self):
if self._host == 'localhost':
return "GPIO%d" % self._number
else:
return "GPIO%d on %s:%d" % (self._number, self._host, self._port)

@property
def host(self):
return self._host

@property
def port(self):
return self._port

@property
def number(self):
return self._number


def _get_state(self):
if self._pwm:
return (
    self._connection.get_PWM_dutycycle(self._number) /
    self._connection.get_PWM_range(self._number)
)
else:
return bool(self._connection.read(self._number))

def _set_state(self, value):
if self._pwm:
try:
value = int(value * self._connection.get_PWM_range(self._number))
if value != self._connection.get_PWM_dutycycle(self._number):
self._connection.set_PWM_dutycycle(self._number, value)
except pigpio.error:
raise PinInvalidState('invalid state "%s" for pin %r' % (value, self))
elif self.function == 'input':
raise PinSetInput('cannot set state of pin %r' % self)
else:
# write forces pin to OUTPUT, hence the check above
self._connection.write(self._number, bool(value))

def _get_pull(self):
return self._pull

def _set_pull(self, value):
if self.function != 'input':
raise PinFixedPull('cannot set pull on non-input pin %r' % self)
if value != 'up' and self._pi_info.pulled_up('GPIO%d' % self._number):
raise PinFixedPull('%r has a physical pull-up resistor' % self)
try:
self._connection.set_pull_up_down(self._number, self.GPIO_PULL_UPS[value])
self._pull = value
except KeyError:
    raise PinInvalidPull('invalid pull "%s" for pin %r' % (value, self))

def _get_frequency(self):
if self._pwm:
return self._connection.get_PWM_frequency(self._number)
return None

def _set_frequency(self, value):
if not self._pwm and value is not None:
    self._connection.set_PWM_frequency(self._number, value)
self._connection.set_PWM_range(self._number, 10000)
self._connection.set_PWM_dutycycle(self._number, 0)
self._pwm = True
elif self._pwm and value is not None:
    if value != self._connection.get_PWM_frequency(self._number):
self._connection.set_PWM_frequency(self._number, value)
self._connection.set_PWM_range(self._number, 10000)
elif self._pwm and value is None:
    self._connection.write(self._number, 0)
self._pwm = False

def _get_bounce(self):
return None if not self._bounce else self._bounce / 1000000

def _set_bounce(self, value):
if value is None:
    value = 0
elif value < 0:
raise PinInvalidBounce('bounce must be 0 or greater')
self._connection.set_glitch_filter(self._number, int(value * 1000000))

def _get_edges(self):
return self.GPIO_EDGES_NAMES[self._edges]

def _set_edges(self, value):
f = self.when_changed
self.when_changed = None
try:
self._edges = self.GPIO_EDGES[value]
finally:
self.when_changed = f

def _get_when_changed(self):
if self._callback is None:
    return None
return self._callback.callb.func

def _set_when_changed(self, value):
if self._callback is not None:
    self._callback.cancel()
self._callback = None
if value is not None:
    self._callback = self._connection.callback(
        self._number, self._edges,
        lambda gpio, level, tick: value())

*/